cni/pkg/nftables/nftables.go

```diff
@@ -0,0 +1,769 @@
+	//   other host processes can also send traffic from the node IP, and nodes can have multiple IPs making the
+	//	 implementation error-prone.
+	//
+	// - Option-3: Use "meta cgroup". This matches based on the cgroup ID of the originating socket.
+	//   However, it only works in the output hook, not in postrouting hook where SNAT can be done.
+	//	 Additionally, cgroup IDs change whenever kubelet restarts, making the solution fragile,
+	//	 and on cgroup v2 clusters it didn’t work properly while testing.
+	//
+	// - Option-4: Use "socket cgroupv2". This supports filtering traffic by the cgroup of the socket
+	//   and works reliably with kernel >= 5.9 and nft >= 0.9.8. However, it is only usable in the output
+	//   hook, not in postrouting, so we cannot directly perform the SNAT operation using a single rule.
+	//	 Also, this works only on cgroup v2 systems.
+	//
+	// - Option-5: Combine "socket cgroupv2" in the output hook with packet marking, then perform SNAT in
+	//   postrouting hook based on the mark. This approach works, but has two downsides:
+	//   (a) it only works on cgroup v2 systems (not cgroup v1), and
+	//   (b) marking all kubelet traffic risks conflicting with CNIs or other components that also rely on packet marks.
+	//
+	//	So, we are using Option-1 (UID-based matching) with "meta skuid <kubelet_uid>". This rule allows
+	//  kubelet-generated health probes while excluding any other host traffic such as kube-proxy.
+	//	It is portable across cgroup versions and k8s distributions. Also, it's slightly stricter than
+	//  iptables "--socket-exists" match which only checks for host-originated sockets.
+
+	cfg.ruleBuilder.AppendRule(PostroutingChain, AmbientNatTable, "meta l4proto tcp", "skuid", kubeletUID,
+		"ip", "daddr", fmt.Sprintf("@%s-v4", config.ProbeIPSet), Counter, "snat", "to", cfg.cfg.HostProbeSNATAddress.String())
+
+	// For V6 we have to use a different set and a different SNAT IP
+	cfg.ruleBuilder.AppendRule(PostroutingChain, AmbientNatTable, "meta l4proto tcp", "skuid", kubeletUID,
+		"ip6", "daddr", fmt.Sprintf("@%s-v6", config.ProbeIPSet), Counter, "snat", "to", cfg.cfg.HostProbeV6SNATAddress.String())
+
+	return util.RunAsHost(func() error {
+		tx, err := cfg.executeHostCommands()
+		if err != nil {
+			log.Errorf("failed to program nftable rules in the host network namespace: %v", err)
+			return err
+		}
+		builder.LogNftRules(tx)
+		return nil
+	})
+}
+
+// DeleteHostRules removes host-level nftables rules
+func (cfg *NftablesConfigurator) DeleteHostRules() {
+	log := log.WithLabels("component", "host")
+	log.Debug("Attempting to delete hostside nftables rules")
+
+	nft, err := cfg.nftProvider("", "")
+	if err != nil {
+		log.Errorf("error creating the nftProvider while deleting the host rules: %v", err)
+		return
+	}
+
+	tx := nft.NewTransaction()
+
+	// TODO: REVISIT: Ensure that the table exists, so that delete does not return any error.
```

Comment:
Not sure if there is the need to revisit it.
The two operations are anyway executed in the same atomic transaction. Add+flush seems like a simple and effective way to handle the deletion.

Reply:
Sorry, it was a stale comment. I'll remove it.